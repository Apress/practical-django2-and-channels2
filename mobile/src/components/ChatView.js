import React from 'react'
import { StyleSheet, Text, TextInput, ScrollView, View } from 'react-native'

const styles = StyleSheet.create({
  chatContainer: {
    borderColor: 'orange',
    borderWidth: 1,
    flex: 1
  },
  chatMessages: {
    borderColor: 'purple',
    borderWidth: 1,
    flex: 1
  },
  chatInput: {
    height: 40,
    margin: 10,
    borderWidth: 1
  },
  chatMessage: {
    backgroundColor: 'lightgrey',
    margin: 10,
    flex: 1,
    flexDirection: 'row',
    alignSelf: 'stretch'
  },
  chatMessageText: {
    flex: 1
  },
  chatMessageType: {
    flex: 1,
    textAlign: 'right'
  }
})

export default class ChatView extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      messages: [],
      shipmentStatus: "n/a"
    }
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  componentDidUpdate (prevProps) {
    if (prevProps.orderCurrentId !== this.props.orderCurrentId) {
      this.props.backendApi.closeMessagesStream()
      this.state = {
        messages: []
      }
      this.props.backendApi
        .fetchShipmentStatus(this.props.orderCurrentId)
        .then(result => {
            this.setState({shipmentStatus: result})
        })
      this.props.backendApi.openMessagesStream(
        this.props.orderCurrentId,
        message => {
          this.setState({
            messages: this.state.messages.concat([message])
          })
        }
      )
    }
  }

  handleSubmit (event) {
    var text = event.nativeEvent.text
    this.props.backendApi.sendMessage(text)
    this.refs.textInput.setNativeProps({ text: '' })
  }

  render () {
    if (this.props.orderCurrentId) {
      return (
        <View style={styles.chatContainer}>
          <Text>Chat for order: {this.props.orderCurrentId}
          - {this.state.shipmentStatus}</Text>
          <ScrollView
            style={styles.chatMessages}
            showsVerticalScrollIndicator={true}
            ref={ref => (this.scrollView = ref)}
            onContentSizeChange={(contentWidth, contentHeight) => {
              this.scrollView.scrollToEnd({ animated: true })
            }}
          >
            {this.state.messages.map((m, index) => {
              return (
                <View key={index} style={styles.chatMessage}>
                  <Text style={styles.chatMessageText}>
                    {m.username} {m.message}
                  </Text>
                  <Text style={styles.chatMessageType}>{m.type}</Text>
                </View>
              )
            })}
          </ScrollView>
          <TextInput
            style={styles.chatInput}
            ref="textInput"
            placeholder="Enter a message..."
            returnKeyType="send"
            onSubmitEditing={this.handleSubmit}
          />
        </View>
      )
    } else {
      return (
        <View>
          <Text>No order selected for chat</Text>
        </View>
      )
    }
  }
}
