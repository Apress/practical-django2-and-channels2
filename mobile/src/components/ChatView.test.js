import React from 'react'
import ChatView from './ChatView'

import renderer from 'react-test-renderer'

it('renders without orders specified', () => {
  const rendered = renderer.create(<ChatView />).toJSON()
  expect(rendered).toBeTruthy()
})

it('renders with orders specified', () => {
  const backendApi = {
    openMessagesStream: jest.fn(),
    closeMessagesStream: jest.fn(),
    sendMessage: jest.fn(),
    fetchShipmentStatus: jest.fn(
      () => new Promise((resolve, reject) => resolve())
    ),
  }
  const rendered = renderer.create(
    <ChatView backendApi={backendApi} orderCurrentId="1" />
  )
  rendered.root.instance.componentDidUpdate({})
  expect(backendApi.openMessagesStream.mock.calls.length).toBe(1)
  expect(backendApi.openMessagesStream.mock.calls[0][0]).toBe('1')
  expect(backendApi.fetchShipmentStatus.mock.calls.length).toBe(1)
  expect(backendApi.fetchShipmentStatus.mock.calls[0][0]).toBe('1')
  rendered.getInstance().setState({shipmentStatus: 'shipped'})
  backendApi.openMessagesStream.mock.calls[0][1]('initial message')
  expect(rendered.toJSON()).toMatchSnapshot()

  rendered.root.instance.handleSubmit({ nativeEvent: { text: 'answer back' } })
  expect(backendApi.sendMessage.mock.calls.length).toBe(1)
  expect(backendApi.sendMessage.mock.calls[0][0]).toBe('answer back')
  expect(rendered.toJSON()).toMatchSnapshot()
})
