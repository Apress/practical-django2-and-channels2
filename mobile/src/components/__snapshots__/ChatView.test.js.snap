// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with orders specified 1`] = `
<View
  style={
    Object {
      "borderColor": "orange",
      "borderWidth": 1,
      "flex": 1,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Chat for order: 
    1
  </Text>
  <RCTScrollView
    onContentSizeChange={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "borderColor": "purple",
        "borderWidth": 1,
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "lightgrey",
            "flex": 1,
            "flexDirection": "row",
            "margin": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
            }
          }
        >
           
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <TextInput
    allowFontScaling={true}
    onSubmitEditing={[Function]}
    placeholder="Enter a message..."
    returnKeyType="send"
    style={
      Object {
        "borderWidth": 1,
        "height": 40,
        "margin": 10,
      }
    }
  />
</View>
`;

exports[`renders with orders specified 2`] = `
<View
  style={
    Object {
      "borderColor": "orange",
      "borderWidth": 1,
      "flex": 1,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Chat for order: 
    1
  </Text>
  <RCTScrollView
    onContentSizeChange={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "borderColor": "purple",
        "borderWidth": 1,
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "lightgrey",
            "flex": 1,
            "flexDirection": "row",
            "margin": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
            }
          }
        >
           
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "flex": 1,
              "textAlign": "right",
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <TextInput
    allowFontScaling={true}
    onSubmitEditing={[Function]}
    placeholder="Enter a message..."
    returnKeyType="send"
    style={
      Object {
        "borderWidth": 1,
        "height": 40,
        "margin": 10,
      }
    }
  />
</View>
`;
